app/src/main/java/org/mozilla/fenix/trackingprotection/TrackingProtectionPanelView.kt

```diff
@@ -0,0 +1,281 @@
+     * @param categoryBlocked The trackers from this category were blocked
+     */
+    fun openDetails(category: TrackingProtectionCategory, categoryBlocked: Boolean)
+}
+
+/**
+ * View that contains and configures the Tracking Protection Panel
+ */
+class TrackingProtectionPanelView(
+    override val containerView: ViewGroup,
+    val interactor: TrackingProtectionPanelInteractor
+) : LayoutContainer {
+    val view: ConstraintLayout = LayoutInflater.from(containerView.context)
+        .inflate(R.layout.component_tracking_protection_panel, containerView, true)
+        .findViewById(R.id.panel_wrapper)
+
+    private val context get() = view.context
+
+    var mode: TrackingProtectionState.Mode = TrackingProtectionState.Mode.Normal
+        private set
+
+    var trackers: List<Tracker> = listOf()
+        private set
+
+    var bucketedTrackers: HashMap<TrackingProtectionCategory, List<String>> = HashMap()
+
+    var loadedTrackers: List<Tracker> = listOf()
+        private set
+
+    var bucketedLoadedTrackers: HashMap<TrackingProtectionCategory, List<String>> = HashMap()
+
+    fun update(state: TrackingProtectionState) {
+        if (state.mode != mode) {
+            mode = state.mode
+        }
+
+        if (state.listTrackers != trackers) {
+            trackers = state.listTrackers
+            bucketedTrackers = getHashMapOfTrackersForCategory(state.listTrackers)
+        }
+
+        if (state.listTrackersLoaded != loadedTrackers) {
+            loadedTrackers = state.listTrackersLoaded
+            bucketedLoadedTrackers = getHashMapOfTrackersForCategory(state.listTrackersLoaded)
+        }
+
+        when (val mode = state.mode) {
+            is TrackingProtectionState.Mode.Normal -> setUIForNormalMode(state)
+            is TrackingProtectionState.Mode.Details -> setUIForDetailsMode(
+                mode.selectedCategory,
+                mode.categoryBlocked
+            )
+        }
+    }
+
+    private fun setUIForNormalMode(state: TrackingProtectionState) {
+        details_mode.visibility = View.GONE
+        normal_mode.visibility = View.VISIBLE
+
+        not_blocking_header.visibility =
+            if (bucketedLoadedTrackers.size == 0) View.GONE else View.VISIBLE
+        bindUrl(state.url)
+        bindTrackingProtectionInfo(state.isTrackingProtectionEnabled)
+        protection_settings.setOnClickListener {
+            interactor.selectTrackingProtectionSettings()
+        }
+
+        blocking_header.visibility =
+            if (bucketedTrackers.size == 0) View.GONE else View.VISIBLE
+        updateCategoryVisibility()
+        setCategoryClickListeners()
+    }
+
+    @Suppress("ComplexMethod")
+    private fun updateCategoryVisibility() {
+        cross_site_tracking.visibility =
+            if (bucketedTrackers[CROSS_SITE_TRACKING_COOKIES]?.isNotEmpty() == true)
+                View.VISIBLE else View.GONE
+        social_media_trackers.visibility =
```

Comment:
This can be improve using a extension function

 ```kotlin
   private fun HashMap<TrackingProtectionCategory, List<String>>.getVisibility(category: TrackingProtectionCategory): Int {
        return if (this[category]?.isNotEmpty() == true)
            View.VISIBLE else View.GONE
    }
```

Reply:
Added a fix for it